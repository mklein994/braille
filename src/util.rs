/// Determine the terminal size from the terminal itself if possible, with fallbacks
pub fn get_terminal_size() -> anyhow::Result<(u16, u16)> {
    use terminal_size::{Height, Width};

    if let Some((Width(width), Height(height))) = terminal_size::terminal_size() {
        Ok((width, height))
    } else {
        use std::env::VarError;

        let parse_from_environment = |name, fallback| match std::env::var(name) {
            Ok(value) => Ok(value.parse()?),
            Err(VarError::NotPresent) => Ok(fallback),
            Err(err) => Err(anyhow::Error::from(err)),
        };

        let width = parse_from_environment("COLUMNS", 80)?;
        let height = parse_from_environment("LINES", 24)?;
        Ok((width, height))
    }
}

/// Scale a value from one range into another
#[must_use]
pub fn scale(value: f64, i_min: f64, i_max: f64, f_min: f64, f_max: f64) -> f64 {
    debug_assert!(i_min < i_max, "i_min < i_max failed: {i_min} < {i_max}");
    debug_assert!(f_min < f_max, "f_min < f_max failed: {f_min} < {f_max}");
    debug_assert!(
        value >= i_min && value <= i_max,
        "value out of bounds: {value} [{i_min}, {i_max}]"
    );

    let slope = (f_max - f_min) / (i_max - i_min);
    f_min + (slope * (value - i_min))
}
