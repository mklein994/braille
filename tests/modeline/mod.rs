use super::util::*;

macro_rules! t_modeline {
    ($name:ident, $input:literal) => {
        #[test]
        fn $name() {
            let (stdout, stderr) = get_output_from_str($input, ["-m".to_string()]);
            insta::assert_snapshot!(stdout);
            assert!(stderr.is_empty());
        }
    };

    (invalid $name:ident, $input:literal) => {
        #[test]
        fn $name() {
            let stuff = $input;
            eprintln!("{stuff:?}");
            let (stdout, stderr) = get_output_from_str($input, ["-m".to_string()]);
            insta::assert_snapshot!(concat!(stringify!($name), "_stdout"), stdout);
            insta::assert_snapshot!(concat!(stringify!($name), "_stderr"), stderr);
        }
    };
}

t_modeline!(
    test_correct_modeline,
    r"braille -r -3 4 4
-2
0
1

3
4"
);

t_modeline!(invalid test_invalid_modeline_prefix, r"braile -r -3 4 4
1
2
3
");

t_modeline!(invalid test_invalid_modeline_arg, r"braille --foo
1
2
3
");
