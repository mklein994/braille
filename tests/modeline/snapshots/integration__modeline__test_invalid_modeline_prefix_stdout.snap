---
source: tests/modeline/mod.rs
expression: stdout
---

